proposals/p2006.md

```diff
@@ -264,9 +265,19 @@ object and address for value expressions when necessary for interop. This
 mechanism may also be used as an escape hatch for value expressions, but that is
 [currently not proposed](#value-expression-escape-hatches).
 
-This interop is based on the fact that these constructs in C++ can bind to
-`const`-declared objects which strongly pressure code to be compatible even with
-Carbon's more restrictive value expression semantics.
+It may seem risky to expose Carbon's value expressions to `const &` parameters
+in this way, as C++ allows casting away `const`. However, in the absence of
+`mutable` members, casting away `const` does not make it safe to _mutate_
+through a `const &` parameter (or a `const`-qualified method). C++ objects can
+be _declared_ `const`, and those objects are allowed to be implemented in
+readonly memory and assumed to not mutate even if escaped. And `const &`
+parameters can bind to these `const`-declared objects. As a consequence,
+actually mutating through a `const &` parameter (perhaps after casting away
+const) is a fundamentally unsafe operation in C++, and one that
+[breaks in real implementations](https://cpp.compiler-explorer.com/z/KMhTondaK).
+As a consequence, there is strong pressure that C++ code operate with `const &`
+parameters in a way that will be fully compatible with Carbon's value
+expressions.
```

Comment:
The case I'm familiar with of mutating a value after casting away `const` is in a situation where you start with a non-`const` object, call a `const` method that returned a `const` reference to some sub-object, cast that `const` away, and then mutate. Basically: casting away `const` that was added by the code in question in order to treat a `const` method as polymorphic on `const`. I'm guessing this is a pattern that we won't support, but we don't expect it to be an issue at the C++ interop boundary?

Reply:
Accepted all the edits, so I think this is good now.